File: VB/dxSampleGroupSort/SortReportTableHelpers/ReportColumnSortHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms
Imports DevExpress.XtraBars
Imports DevExpress.XtraReports.UI
Imports System.Linq
Imports DevExpress.Utils.Win

Namespace ReportSortHelpers
    Public Class ReportColumnSortHelper
        Private Menu As PopupMenu
        Private _sortPrefix As String = "Sort_"
        Private ClickedItem As HelpSortItem
        Private sortItems As List(Of HelpSortItem)
        Friend Property Report() As XtraReport

        Public Sub New(ByVal report As XtraReport)
            Me.Report = report
            Menu = Menu_InitSortMenu()
            AddHandler Menu.Popup, AddressOf Menu_Popup

            Dim sortCells As New List(Of XRLabel)()
            sortItems = New List(Of HelpSortItem)()
            InitHeaderCells(sortCells)
        End Sub

        Private Sub Menu_Popup(ByVal sender As Object, ByVal e As EventArgs)
            Dim popupcontrol As IPopupControl = TryCast(sender, IPopupControl)
        End Sub
        Private Function Menu_InitSortMenu() As PopupMenu
            Dim manager As New BarManager()
            Dim sortMenu As New PopupMenu()
            AddHandler manager.ItemClick, AddressOf manager_ItemClick
            sortMenu.Manager = manager
            Dim item1 As BarCheckItem = Menu_InitSortItem(SortOrder.None)
            Dim item2 As BarCheckItem = Menu_InitSortItem(SortOrder.Ascending)
            Dim item3 As BarCheckItem = Menu_InitSortItem(SortOrder.Descending)
            item3.GroupIndex = 1
            item2.GroupIndex = item3.GroupIndex
            item1.GroupIndex = item2.GroupIndex
            sortMenu.AddItems(New BarItem() { item1, item2, item3 })
            Return sortMenu
        End Function
        Private Shared Function Menu_InitSortItem(ByVal sortMode As SortOrder) As BarCheckItem
            Dim itemNone As New BarCheckItem()
            itemNone.Width = 100
            itemNone.Tag = sortMode
            itemNone.Caption = itemNone.Tag.ToString()
            itemNone.CheckBoxVisibility = CheckBoxVisibility.BeforeText
            Return itemNone
        End Function

        Private Sub InitHeaderCells(ByVal sortCells As List(Of XRLabel))
            PopulateHeaderCells(sortCells)
            For Each lbl As XRLabel In sortCells
                Dim fieldName As String = lbl.Tag.ToString().Substring(lbl.Tag.ToString().IndexOf("_"c) + 1)
                Dim caption As String = lbl.Text
                Dim item As New HelpSortItem(Me, lbl, Menu, fieldName, caption)
                sortItems.Add(item)
                AddHandler lbl.PreviewMouseDown, AddressOf lbl_PreviewMouseDown
            Next lbl
        End Sub
        Private Sub lbl_PreviewMouseDown(ByVal sender As Object, ByVal e As PreviewMouseEventArgs)
            Dim clickedCell As XRLabel = TryCast(sender, XRLabel)
            ClickedItem = CType(sortItems.FirstOrDefault(Function(x) x.HeaderCell Is clickedCell), HelpSortItem)
        End Sub
        Private Sub PopulateHeaderCells(ByVal sortCells As List(Of XRLabel))
            Dim allControls As IEnumerable(Of XRLabel) = GetAllReportControls()
            Dim cells = From c In allControls _
                        Where c.Tag IsNot Nothing AndAlso c.Tag.ToString().StartsWith(_sortPrefix) _
                        Select c
            sortCells.AddRange(cells.ToList())
        End Sub
        Private Function GetAllReportControls() As IEnumerable(Of XRLabel)
            Return Report.AllControls(Of XRLabel)()
        End Function

        Private Sub manager_ItemClick(ByVal sender As Object, ByVal e As ItemClickEventArgs)
            Dim clickedMenuItem As BarItem = e.Item
            Dim item As HelpSortItem = Nothing
            If clickedMenuItem IsNot Nothing AndAlso clickedMenuItem.Tag IsNot Nothing AndAlso TypeOf clickedMenuItem.Tag Is SortOrder Then
                item = ClickedItem
                Dim newSortOrder As SortOrder = CType(clickedMenuItem.Tag, SortOrder)
                item.ColumnSortOrder = newSortOrder
            End If
            If item IsNot Nothing AndAlso item.Modified Then
                RecreateDocument()
            End If
        End Sub
        Friend Sub RecreateDocument()
            Report.CreateDocument()
        End Sub
        Public Function IsMouseButtonPressed(ByVal button As MouseButtons) As Boolean
            Return (Control.MouseButtons And button) = button
        End Function
        Friend Sub ResetItemsCaption(ByVal actualItem As HelpSortItem)
            Dim order As SortOrder = SortOrder.None
            For Each item As HelpSortItem In sortItems
                If actualItem IsNot Nothing AndAlso item Is actualItem Then
                    Continue For
                Else
                    item.ColumnSortOrder = order
                End If
            Next item
        End Sub
        Private Sub ClearSorting()
            Dim dataBand As DetailBand = GetDataBand()
            dataBand.SortFields.Clear()
        End Sub
        Private Function GetDataBand() As DetailBand
            Return CType(Report.Bands.FirstOrDefault(Function(x) TypeOf x Is DetailBand), DetailBand)
        End Function
        Public Sub ClearAll(ByVal actualItem As HelpSortItem)
            ClearAll(actualItem, True)
        End Sub
        Public Sub ClearAll(ByVal actualItem As HelpSortItem, ByVal refreshDocument As Boolean)
            ClearSorting()
            ResetItemsCaption(actualItem)
            If refreshDocument Then
                RecreateDocument()
            End If
        End Sub
        Public Sub HidePopup()
            If Menu.Opened Then
                Menu.HidePopup()
            End If

        End Sub
    End Class
End Namespace

File: CS/dxSampleGroupSort/SortReportTableHelpers/ReportColumnSortHelper.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;
using DevExpress.XtraBars;
using DevExpress.XtraReports.UI;
using System.Linq;
using DevExpress.Utils.Win;
using DevExpress.XtraPrinting.Control;

namespace ReportSortHelpers {
    public class ReportColumnSortHelper {        
        private string _sortPrefix = "Sort_";
        private string _clearSort = "Clear_Sort";
        private HelpSortItem ClickedItem;
        private List<HelpSortItem> sortItems;
        private BarManager manager;
        internal PopupMenu Menu;
        internal XtraReport Report { get; set; }
        internal ReportPrintTool PrintTool { get; set; }

        internal BarManager BarManager
        {
            get { return manager; }
            set
            {
                manager = value;
                if (manager != null)
                    Menu = Menu_InitSortMenu();
            }
        }

        public ReportColumnSortHelper(XtraReport report) {
            Report = report;
            PrintTool = new ReportPrintTool(report);            

            List<XRLabel> sortCells = new List<XRLabel>();
            sortItems = new List<HelpSortItem>();
            InitHeaderCells(sortCells);
            InitClearSortingCell();
        }        

        #region PreviewMethods
        public void ShowPreview()
        {
            BarManager = PrintTool.PreviewForm.PrintBarManager;
            PrintTool.ShowPreview();            
        }

        public void ShowPreviewDialog()
        {
            BarManager = PrintTool.PreviewForm.PrintBarManager;
            PrintTool.ShowPreviewDialog();
        }

        public void ShowRibbonPreview()
        {
            BarManager = PrintTool.PreviewRibbonForm.Ribbon.Manager;
            PrintTool.ShowRibbonPreview();
        }

        public void ShowRibbonPreviewDialog()
        {
            BarManager = PrintTool.PreviewRibbonForm.Ribbon.Manager;
            PrintTool.ShowRibbonPreviewDialog();
        }
        #endregion

        #region Clear_Sorting
        private void InitClearSortingCell()
        {
            XRLabel clearLabel = Report.AllControls<XRLabel>().FirstOrDefault(x => Convert.ToString(x.Tag).Equals(_clearSort));
            if (clearLabel != null)
            {
                clearLabel.PreviewMouseDown += clearLabel_PreviewMouseDown;
                clearLabel.PreviewMouseMove += clearLabel_PreviewMouseMove;
            }
        }

        void clearLabel_PreviewMouseMove(object sender, PreviewMouseEventArgs e)
        {
            SetHandCursor(e);
        }

        void clearLabel_PreviewMouseDown(object sender, PreviewMouseEventArgs e)
        {
            ClearAll(null);
        }

        private static void SetHandCursor(PreviewMouseEventArgs e)
        {
            e.PreviewControl.Cursor = Cursors.Hand;
        }
        #endregion

        #region Header Cells
        private void InitHeaderCells(List<XRLabel> sortCells)
        {
            PopulateHeaderCells(sortCells);
            foreach (XRLabel lbl in sortCells)
            {
                string fieldName = lbl.Tag.ToString().Substring(lbl.Tag.ToString().IndexOf('_') + 1);
                string caption = lbl.Text;
                HelpSortItem item = new HelpSortItem(this, lbl, Menu, fieldName, caption);
                sortItems.Add(item);
                lbl.PreviewMouseDown += lbl_PreviewMouseDown;
            }
        }
        private void lbl_PreviewMouseDown(object sender, PreviewMouseEventArgs e)
        {
            XRLabel clickedCell = sender as XRLabel;
            ClickedItem = (HelpSortItem)sortItems.FirstOrDefault(x => x.HeaderCell == clickedCell);       
        }

        private void PopulateHeaderCells(List<XRLabel> sortCells)
        {
            IEnumerable<XRLabel> allControls = GetAllReportControls();
            var cells = from c in allControls
                        where c.Tag != null && c.Tag.ToString().StartsWith(_sortPrefix)
                        select c;
            sortCells.AddRange(cells.ToList());
        }
        private IEnumerable<XRLabel> GetAllReportControls()
        {
            return Report.AllControls<XRLabel>();
        }
        #endregion

        #region Menu initialization
        private PopupMenu Menu_InitSortMenu()
        {
            PopupMenu sortMenu = new PopupMenu();
            manager.ItemClick += manager_ItemClick;
            sortMenu.Manager = BarManager;
            BarCheckItem item1 = Menu_InitSortItem(SortOrder.None);
            BarCheckItem item2 = Menu_InitSortItem(SortOrder.Ascending);
            BarCheckItem item3 = Menu_InitSortItem(SortOrder.Descending);
            item1.GroupIndex = item2.GroupIndex = item3.GroupIndex = 1;
            sortMenu.AddItems(new BarItem[] { item1, item2, item3 });
            return sortMenu;        
        }

        private static BarCheckItem Menu_InitSortItem(SortOrder sortMode)
        {
            BarCheckItem itemNone = new BarCheckItem();
            itemNone.Width = 100;
            itemNone.Tag = sortMode;
            itemNone.Caption = itemNone.Tag.ToString();
            itemNone.CheckBoxVisibility = CheckBoxVisibility.BeforeText;
            return itemNone;
        }

        private void manager_ItemClick(object sender, ItemClickEventArgs e)
        {
            BarItem clickedMenuItem = e.Item;
            HelpSortItem item = null;
            if (clickedMenuItem != null && clickedMenuItem.Tag != null && clickedMenuItem.Tag is SortOrder)
            {
                item = ClickedItem;
                SortOrder newSortOrder = (SortOrder)clickedMenuItem.Tag;
                item.ColumnSortOrder = newSortOrder;
            }
            if (item != null && item.Modified)
                RecreateDocument();
        }
        #endregion        

        internal void RecreateDocument() {
            Report.CreateDocument();
        }
        public bool IsMouseButtonPressed(MouseButtons button) {
            return (Control.MouseButtons & button) == button;
        }
        internal void ResetItemsCaption(HelpSortItem actualItem) {
            SortOrder order = SortOrder.None;
            foreach (HelpSortItem item in sortItems) {
                if (actualItem != null && item == actualItem) {
                    continue;
                } else {
                    item.ColumnSortOrder = order;
                }
            }
        }
        private void ClearSorting() {
            DetailBand dataBand = GetDataBand();
            dataBand.SortFields.Clear();
        }
        private DetailBand GetDataBand() {
            return (DetailBand)Report.Bands.FirstOrDefault(x => x is DetailBand);
        }
        public void ClearAll(HelpSortItem actualItem) {
            ClearAll(actualItem, true);
        }
        public void ClearAll(HelpSortItem actualItem, bool refreshDocument) {
            ClearSorting();
            ResetItemsCaption(actualItem);
            if(refreshDocument)
                RecreateDocument();
        }
        public void HidePopup() {
            if (Menu.Opened) {
                Menu.HidePopup();
            }

        }
    }
}
